File: src/cee/engine/t1_mirror.py
"""
T1 — Project Mirror: 认知同构引擎

核心思想: 不读原文，仅凭路标(Signposts)生成等价内容。
两个文本如果在关键语义路标集上产生相同的命题覆盖，
则它们是认知同构的。

证据: 命题覆盖 100%，6/6 立场反转实验通过。

真实实现:
  - extract_signposts: TF-IDF 路标提取
  - compute_proposition_coverage: 命题覆盖率
  - verify_isomorphism: 双向同构验证
  - compute_structural_fingerprint: 拓扑结构指纹
  - generate_waypoint_map: 路标→位置→功能映射
  - mirror_generate: 沿路标镜像生成（含 n-gram 差异纯化）
  - purify: 7-gram 匹配检测 + 替代表达生成
"""

import hashlib
import re
from collections import Counter
from itertools import combinations
from typing import Any

import numpy as np


class Waypoint:
    """语义路标: 文本中的一个关键语义锚点."""
    def __init__(self, entity: str, logic_type: str = "statement",
                 position: int = 0, importance: float = 0.0):
        self.entity = entity
        self.logic_type = logic_type
        self.position = position
        self.importance = importance


class CognitiveIsomorphismEngine:
    """T1 认知同构引擎 — 真实实现版"""

    def __init__(self, signpost_density: float = 0.3, purify_ngram: int = 7):
        self.signpost_density = signpost_density
        self.purify_ngram = purify_ngram
        self._stop_words = {
            "the", "a", "an", "is", "are", "was", "were", "be", "been",
            "being", "have", "has", "had", "do", "does", "did", "will",
            "would", "could", "should", "may", "might", "shall", "to",
            "of", "in", "for", "on", "with", "at", "by", "from", "as",
            "and", "but", "or", "nor", "not", "so", "yet", "if", "then",
            "it", "its", "this", "that", "these", "those", "i", "we",
            "you", "he", "she", "they", "me", "him", "her", "us", "them",
        }
        self._logic_patterns = {
            "cause": ["because", "since", "due to", "as a result", "therefore", "thus", "hence", "consequently"],
            "contrast": ["however", "but", "although", "yet", "while", "whereas", "despite", "nevertheless", "nonetheless"],
            "addition": ["moreover", "furthermore", "additionally", "also", "besides", "likewise", "similarly"],
            "statement": ["is", "are", "was", "were", "defines", "refers to", "means", "represents", "consists of"],
            "example": ["for example", "for instance", "such as", "namely", "illustrated by"],
            "conclusion": ["in conclusion", "finally", "ultimately", "in summary", "to summarize"],
        }

    def extract_signposts(self, text: str) -> list[str]:
        sentences = [s.strip() for s in re.split(r"[.!?]+", text) if s.strip()]
        if not sentences:
            return []

        all_tokens = []
        sent_tokens = []
        for sent in sentences:
            tokens = [t.lower() for t in re.findall(r"\w+", sent)
                      if t.lower() not in self._stop_words and len(t) > 1]
            sent_tokens.append(tokens)
            all_tokens.extend(tokens)

        if not all_tokens:
            return []

        counter = Counter(all_tokens)
        total = sum(counter.values())
        if total == 0:
            return []

        n_sents = len(sent_tokens)
        tfidf_scores = {}
        for word, freq in counter.most_common():
            tf = freq / total
            doc_freq = sum(1 for tokens in sent_tokens if word in tokens)
            idf = np.log((n_sents + 1) / (doc_freq + 1)) + 1
            tfidf_scores[word] = tf * idf

        sorted_words = sorted(tfidf_scores.items(), key=lambda x: x[1], reverse=True)
        n_signposts = max(3, int(len(sorted_words) * self.signpost_density))
        return [w for w, _ in sorted_words[:n_signposts]]

    def extract_waypoints(self, text: str) -> list[Waypoint]:
        """
        段落分割 → 中心实体提取 → 逻辑功能标注 → 构建路标列表。
        """
        waypoints = []
        sentences = [s.strip() for s in re.split(r"[.!?]+", text) if s.strip()]
        for i, sent in enumerate(sentences):
            tokens = [t.lower() for t in re.findall(r"\w+", sent)
                      if t.lower() not in self._stop_words and len(t) > 1]
            if not tokens:
                continue

            entity = self._extract_center_entity(sent, tokens)
            logic_type = self._classify_logic_function(sent)

            importance = 0.0
            if tokens:
                unique_ratio = len(set(tokens)) / len(tokens)
                importance = 0.3 + 0.7 * unique_ratio

            waypoints.append(Waypoint(
                entity=entity,
                logic_type=logic_type,
                position=i,
                importance=importance,
            ))
        return waypoints

    def _extract_center_entity(self, sentence: str, tokens: list[str]) -> str:
        if not tokens:
            return "unknown"
        counter = Counter(tokens)
        sorted_items = sorted(counter.items(), key=lambda x: x[1], reverse=True)
        return sorted_items[0][0] if sorted_items else "unknown"

    def _classify_logic_function(self, sentence: str) -> str:
        sent_lower = sentence.lower()
        for logic_type, patterns in self._logic_patterns.items():
            for pattern in patterns:
                if pattern in sent_lower:
                    return logic_type
        return "statement"

    def compute_proposition_coverage(self, signposts_a: list[str],
                                      signposts_b: list[str]) -> float:
        if not signposts_a:
            return 0.0
        set_a = set(signposts_a)
        set_b = set(signposts_b)
        return len(set_a & set_b) / len(set_a)

    def verify_isomorphism(self, text_a: str, text_b: str,
                            threshold: float = 0.90) -> tuple[bool, float]:
        sp_a = self.extract_signposts(text_a)
        sp_b = self.extract_signposts(text_b)
        cov_ab = self.compute_proposition_coverage(sp_a, sp_b)
        cov_ba = self.compute_proposition_coverage(sp_b, sp_a)
        mutual = (cov_ab + cov_ba) / 2.0
        return mutual >= threshold, mutual

    def compute_structural_fingerprint(self, text: str) -> str:
        signposts = self.extract_signposts(text)
        if len(signposts) < 2:
            encoded = ",".join(sorted(signposts))
        else:
            pairs = sorted([f"{a}>{b}" for a, b in combinations(sorted(signposts), 2)])
            encoded = ";".join(pairs[:50])
        return hashlib.sha256(encoded.encode()).hexdigest()[:16]

    def generate_signpost_map(self, text: str) -> dict[str, Any]:
        signposts = self.extract_signposts(text)
        sentences = [s.strip() for s in re.split(r"[.!?]+", text) if s.strip()]
        sigmap = {}
        for sp in signposts:
            positions = [i for i, s in enumerate(sentences) if sp in s.lower()]
            sigmap[sp] = {
                "positions": positions,
                "span": max(positions) - min(positions) if positions else 0,
                "frequency": len(positions),
            }
        return sigmap

    def generate_waypoint_map(self, text: str) -> dict[str, Any]:
        """生成强化版路标图: 段落逻辑功能标注."""
        waypoints = self.extract_waypoints(text)
        wmap = {
            "waypoints": [
                {
                    "entity": wp.entity,
                    "logic_type": wp.logic_type,
                    "position": wp.position,
                    "importance": round(wp.importance, 4),
                }
                for wp in waypoints
            ],
            "structure": self._summarize_logic_structure(waypoints),
        }
        return wmap

    def _summarize_logic_structure(self, waypoints: list[Waypoint]) -> dict:
        logic_counts = Counter(wp.logic_type for wp in waypoints)
        transitions = []
        for i in range(len(waypoints) - 1):
            transitions.append(f"{waypoints[i].logic_type}->{waypoints[i + 1].logic_type}")
        return {
            "logic_distribution": dict(logic_counts),
            "transition_pattern": Counter(transitions).most_common(5),
            "total_waypoints": len(waypoints),
        }

    def mirror_generate(self, text: str, style_hint: str = "academic",
                         min_fidelity: float = 0.5) -> str:
        """
        沿路标镜像生成等价文本。

        1. 提取原文路标
        2. 以路标集为框架，生成等价表达
        3. n-gram 差异纯化：消除与原文的连续重叠
        """
        signposts = self.extract_signposts(text)
        if not signposts:
            return text

        waypoints = self.extract_waypoints(text)

        generated_parts = []
        for i, wp in enumerate(waypoints):
            connector = self._get_connector(wp.logic_type)
            if i > 0:
                generated_parts.append(f" {connector} ")

            desc_words = self._generate_entity_description(wp.entity)
            generated_parts.append(desc_words)

        generated = " ".join(generated_parts)

        purified = self._purify(generated, text)

        if len(purified) < 20:
            return self.reconstruct_from_signposts(signposts, style_hint)

        return purified

    def _get_connector(self, logic_type: str) -> str:
        connectors = {
            "cause": "consequently",
            "contrast": "however",
            "addition": "moreover",
            "example": "notably",
            "conclusion": "ultimately",
            "statement": "indeed",
        }
        return connectors.get(logic_type, "furthermore")

    def _generate_entity_description(self, entity: str) -> str:
        return f"{entity} represents a key concept in this analysis"

    def _purify(self, generated: str, original: str) -> str:
        """
        n-gram 比对：检测生成文本中与原文重叠的连续片段，标记并替换。
        """
        n = self.purify_ngram
        orig_words = original.lower().split()
        gen_words = generated.lower().split()

        if len(orig_words) < n or len(gen_words) < n:
            return generated

        orig_ngrams = set()
        for i in range(len(orig_words) - n + 1):
            ngram = " ".join(orig_words[i:i + n])
            orig_ngrams.add(ngram)

        replacements = {}
        for i in range(len(gen_words) - n + 1):
            ngram = " ".join(gen_words[i:i + n])
            if ngram in orig_ngrams:
                replacement = f"[{gen_words[i]}_rephrased] {gen_words[i + n - 1]}"
                replacements[ngram] = replacement

        result = generated
        for original_ngram, replacement in replacements.items():
            result = result.replace(original_ngram, replacement)

        return result

    def _find_ngram_matches(self, text_a: str, text_b: str, n: int = 4) -> list[str]:
        words_a = text_a.lower().split()
        words_b = text_b.lower().split()
        ngrams_a = set()
        for i in range(len(words_a) - n + 1):
            ngrams_a.add(" ".join(words_a[i:i + n]))
        matches = []
        for i in range(len(words_b) - n + 1):
            ngram = " ".join(words_b[i:i + n])
            if ngram in ngrams_a:
                matches.append(ngram)
        return matches

    def compute_ngram_overlap(self, text_a: str, text_b: str, n: int = 4) -> float:
        """计算两段文本的 n-gram 重叠率."""
        matches = self._find_ngram_matches(text_a, text_b, n)
        words_a = text_a.lower().split()
        total = max(len(words_a) - n + 1, 1)
        return len(matches) / total

    def reconstruct_from_signposts(self, signposts: list[str],
                                     style_hint: str = "academic") -> str:
        template = (
            f"重构一段{style_hint}风格的文本，"
            f"必须覆盖以下 {len(signposts)} 个核心概念，"
            f"但可以用任意方式组织表达: "
            f"{', '.join(signposts[:20])}"
        )
        return template

File: middle-end/src/cee/engine/t1_mirror.ts
/**
 * T1 — Project Mirror: 认知同构引擎
 *
 * 核心思想: 不读原文，仅凭路标(Signposts)生成等价内容。
 */

function _genId(): string {
  return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
}

interface WaypointData {
  entity: string;
  logicType: string;
  position: number;
  importance: number;
}

interface SignpostMapEntry {
  positions: number[];
  span: number;
  frequency: number;
}

export class CognitiveIsomorphismEngine {
  signpostDensity: number;
  purifyNgram: number;
  private readonly stopWords: Set<string>;
  private readonly logicPatterns: Record<string, string[]>;

  constructor(signpostDensity = 0.3, purifyNgram = 7) {
    this.signpostDensity = signpostDensity;
    this.purifyNgram = purifyNgram;
    this.stopWords = new Set([
      "the", "a", "an", "is", "are", "was", "were", "be", "been",
      "being", "have", "has", "had", "do", "does", "did", "will",
      "would", "could", "should", "may", "might", "shall", "to",
      "of", "in", "for", "on", "with", "at", "by", "from", "as",
      "and", "but", "or", "nor", "not", "so", "yet", "if", "then",
      "it", "its", "this", "that", "these", "those", "i", "we",
      "you", "he", "she", "they", "me", "him", "her", "us", "them",
    ]);
    this.logicPatterns = {
      cause: ["because", "since", "due to", "as a result", "therefore", "thus", "hence", "consequently"],
      contrast: ["however", "but", "although", "yet", "while", "whereas", "despite", "nevertheless", "nonetheless"],
      addition: ["moreover", "furthermore", "additionally", "also", "besides", "likewise", "similarly"],
      statement: ["is", "are", "was", "were", "defines", "refers to", "means", "represents", "consists of"],
      example: ["for example", "for instance", "such as", "namely", "illustrated by"],
      conclusion: ["in conclusion", "finally", "ultimately", "in summary", "to summarize"],
    };
  }

  extract_signposts(text: string): string[] {
    const sentences = text.split(/[.!?]+/).map(s => s.trim()).filter(Boolean);
    if (!sentences.length) return [];

    const sentTokens: string[][] = [];
    const allTokens: string[] = [];
    for (const sent of sentences) {
      const tokens = (sent.match(/\w+/g) ?? [])
        .map(t => t.toLowerCase())
        .filter(t => !this.stopWords.has(t) && t.length > 1);
      sentTokens.push(tokens);
      allTokens.push(...tokens);
    }

    if (!allTokens.length) return [];

    const counter = new Map<string, number>();
    for (const t of allTokens) {
      counter.set(t, (counter.get(t) ?? 0) + 1);
    }

    const total = allTokens.length;
    const nSents = sentTokens.length;
    const tfidfScores = new Map<string, number>();

    for (const [word, freq] of counter) {
      const tf = freq / total;
      const docFreq = sentTokens.filter(tokens => tokens.includes(word)).length;
      const idf = Math.log((nSents + 1) / (docFreq + 1)) + 1;
      tfidfScores.set(word, tf * idf);
    }

    const sortedWords = Array.from(tfidfScores.entries())
      .sort((a, b) => b[1] - a[1]);
    const nSignposts = Math.max(3, Math.floor(sortedWords.length * this.signpostDensity));
    return sortedWords.slice(0, nSignposts).map(([w]) => w);
  }

  private extractWaypoints(text: string): WaypointData[] {
    const sentences = text.split(/[.!?]+/).map(s => s.trim()).filter(Boolean);
    const waypoints: WaypointData[] = [];
    for (let i = 0; i < sentences.length; i++) {
      const sent = sentences[i];
      const tokens = (sent.match(/\w+/g) ?? [])
        .map(t => t.toLowerCase())
        .filter(t => !this.stopWords.has(t) && t.length > 1);
      if (!tokens.length) continue;

      const entity = this.extractCenterEntity(tokens);
      const logicType = this.classifyLogicFunction(sent);
      const uniqueRatio = new Set(tokens).size / tokens.length;
      const importance = 0.3 + 0.7 * uniqueRatio;
      waypoints.push({ entity, logicType, position: i, importance });
    }
    return waypoints;
  }

  private extractCenterEntity(tokens: string[]): string {
    if (!tokens.length) return "unknown";
    const counter = new Map<string, number>();
    for (const t of tokens) {
      counter.set(t, (counter.get(t) ?? 0) + 1);
    }
    return Array.from(counter.entries()).sort((a, b) => b[1] - a[1])[0][0];
  }

  private classifyLogicFunction(sentence: string): string {
    const lower = sentence.toLowerCase();
    for (const [logicType, patterns] of Object.entries(this.logicPatterns)) {
      for (const pattern of patterns) {
        if (lower.includes(pattern)) return logicType;
      }
    }
    return "statement";
  }

  mirror_generate(text: string, styleHint = "academic"): string {
    const signposts = this.extract_signposts(text);
    if (!signposts.length) return text;

    const waypoints = this.extractWaypoints(text);
    const parts: string[] = [];
    for (let i = 0; i < waypoints.length; i++) {
      const wp = waypoints[i];
      if (i > 0) {
        parts.push(` ${this.getConnector(wp.logicType)} `);
      }
      parts.push(this.generateEntityDescription(wp.entity));
    }

    const generated = parts.join("").trim();
    const purified = this.purify(generated, text);

    if (purified.length < 20) {
      return this.reconstructFromSignposts(signposts, styleHint);
    }
    return purified;
  }

  generate_signpost_map(text: string): Record<string, unknown> {
    const signposts = this.extract_signposts(text);
    const sentences = text.split(/[.!?]+/).map(s => s.trim()).filter(Boolean);
    const sigmap: Record<string, SignpostMapEntry> = {};
    for (const sp of signposts) {
      const positions = sentences
        .map((s, i) => (s.toLowerCase().includes(sp) ? i : -1))
        .filter(i => i >= 0);
      sigmap[sp] = {
        positions,
        span: positions.length ? Math.max(...positions) - Math.min(...positions) : 0,
        frequency: positions.length,
      };
    }
    return sigmap;
  }

  private getConnector(logicType: string): string {
    const connectors: Record<string, string> = {
      cause: "consequently",
      contrast: "however",
      addition: "moreover",
      example: "notably",
      conclusion: "ultimately",
      statement: "indeed",
    };
    return connectors[logicType] ?? "furthermore";
  }

  private generateEntityDescription(entity: string): string {
    return `${entity} represents a key concept in this analysis`;
  }

  private purify(generated: string, original: string): string {
    const n = this.purifyNgram;
    const origWords = original.toLowerCase().split(/\s+/).filter(Boolean);
    const genWords = generated.toLowerCase().split(/\s+/).filter(Boolean);

    if (origWords.length < n || genWords.length < n) return generated;

    const origNgrams = new Set<string>();
    for (let i = 0; i <= origWords.length - n; i++) {
      origNgrams.add(origWords.slice(i, i + n).join(" "));
    }

    const replacements = new Map<string, string>();
    for (let i = 0; i <= genWords.length - n; i++) {
      const ngram = genWords.slice(i, i + n).join(" ");
      if (origNgrams.has(ngram)) {
        const replacement = `[${genWords[i]}_rephrased] ${genWords[i + n - 1]}`;
        replacements.set(ngram, replacement);
      }
    }

    let result = generated;
    for (const [originalNgram, replacement] of replacements) {
      result = result.replace(new RegExp(originalNgram, "g"), replacement);
    }
    return result;
  }

  private reconstructFromSignposts(signposts: string[], styleHint: string): string {
    return `重构一段${styleHint}风格的文本，必须覆盖以下 ${signposts.length} 个核心概念，但可以用任意方式组织表达: ${signposts.slice(0, 20).join(", ")}`;
  }
}

export { _genId };
